### PR TITLE
extend/os/mac/keg_relocate: implement `RPATH` relocation

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This implements `RPATH` relocation on macOS.

I've tested this successfully locally, but I've gated this behind `HOMEBREW_RELOCATE_RPATHS` for now so that I can test this with various formulae in an external tap when this is merged.

Using a bottled `hello`, built with the appropriate `LDFLAGS` to create an `LC_RPATH` entry:
```
❯ otool -l ./hello | rg -A2 LC_RPATH
          cmd LC_RPATH
      cmdsize 48
         path @@HOMEBREW_PREFIX@@/opt/llvm/lib (offset 12)
```

After pouring the bottle:
```
❯ otool -l $(brew --prefix hello)/bin/hello | rg -A2 LC_RPATH
          cmd LC_RPATH
      cmdsize 40
         path /usr/local/opt/llvm/lib (offset 12)
```

See #11329.

### CODE REVIEW

File: Library/Homebrew/extend/os/mac/keg_relocate.rb [hunk 1]
```diff
@@ -111,6 +123,12 @@ def each_install_name_for(file, &block)
     dylibs.each(&block)
   end
 
+  def each_rpath_for(file, &block)
+    rpaths = file.rpaths
+    rpaths.reject! { |fn| fn =~ /^@(loader|executable)_path/ }
```
Comment (Bo98) [inline Library/Homebrew/extend/os/mac/keg_relocate.rb:128]: I don't understand the need for this. All we do is `start_with?` checks anyway?

**Review by Bo98 — APPROVED** [review]
LGTM. I think this is pretty safe since we're not "losing" anything here but if you feel more comfortable with a `HOMEBREW_RELOCATE_RPATHS` then that's fine.

**Comment by carlocab** [comment]
> LGTM. I think this is pretty safe since we're not "losing" anything here but if you feel more comfortable with a `HOMEBREW_RELOCATE_RPATHS` then that's fine.

Without gating, `brew bottle` will replace `LC_RPATH` prefixes with placeholders upon bottling. This will be a problem for users still on the last release tag, because `brew` won't be replacing those placeholders upon pouring, won't it?

File: Library/Homebrew/extend/os/mac/keg_relocate.rb (hunk 1 above)
Comment (carlocab) [reply Library/Homebrew/extend/os/mac/keg_relocate.rb:128]: This is a good question. I just copied the logic from here:

https://github.com/Homebrew/brew/blob/3e327626079e22f4becebb26b632ad431400105e/Library/Homebrew/extend/os/mac/keg_relocate.rb#L107-L112

I guess the idea was to save some `start_with?` checks, but that's just replacing one operation with another... I can remove it if you prefer. Though we should probably do the same with `each_install_name_for` unless there's a reason those checks belong there but not here.

**Comment by Bo98** [comment]
Yes, same problem with #11286. Needs to be merged and tagged quickly after (maybe with a homebrew-core test run first), or we only implement the `@@HOMEBREW_PREFIX@@` -> real prefix side first.

ENV flag doesn't really solve that though (except for local testing) since end users will also need that flag to process such bottles.

**Comment by carlocab** [comment]
Yup, just wanted this in `brew` first so that I can test builds on GitHub CI rather than my local potato. But if the extra testing seems unnecessary, I don't mind shipping this with a tag without the gating too.

File: Library/Homebrew/extend/os/mac/keg_relocate.rb (hunk 1 above)
Comment (carlocab) [reply Library/Homebrew/extend/os/mac/keg_relocate.rb:128]: Actually, doing `reject!` is probably way less efficient than skipping straight to the `start_with?` checks.

**Comment by Bo98** [comment]
> I can test builds on GitHub CI rather than my local potato

That's fair - but how do you plan to do this? (Just so I'm on the same page here)

**Comment by carlocab** [comment]
I was thinking that we'd merge this, then I'd create a throwaway tap and try to run `test-bot` on some formulae in that tap to check that they're not broken by this.

**Comment by Bo98** [comment]
Ok, I'm fine with that.

When the env flag is removed, we will still have the same merge strategy question as #11286. Might make sense to handle them at the same time, though note that the relocation API has changed in it (I can separate that out if it's easier).

**Comment by carlocab** [comment]
> When the env flag is removed, we will still have the same merge strategy question as #11286. Might make sense to handle them at the same time, though note that the relocation API has changed in the other PR (I can separate that out if it's easier).

I'd forgotten you'd changed the relocation API there. I'd suggest we merge this PR first since this one is gated, then you can rebase #11286 to also adjust the changes I've made here to the new API (I can do that too if you like).

The env flag can also be removed in #11286 so that all the new relocation code can be shipped in the same tag.

**Comment by Bo98** [comment]
That sounds like a good idea.

> I hope this doesn't impose an inconvenient timeline on your PR.

No, it'll be fine I think.

Assuming testing goes well and no issues arise, when do you reckon we should ship #11286? (Hope this doesn't sound like I'm rushing - I'm just trying to plan this a bit.)

The merge will be something like:
* Make sure no bottles are going to be built & published (perhaps pick a time just after automerge runs) during the below process.
* Merge PR.
* Run a test PR - I propose `git` & `subversion` to test `git-svn`.
* Tag a brew release if it goes well (or revert if it doesn't).

**Comment by carlocab** [comment]
> Assuming testing goes well and no issues arise, when do you reckon we should ship #11286?

That depends on when we can get the Intel Big Sur nodes up to 11.3. If possible (but fine too if not), I suggest we do that early this weekend, and then do the #11286 merge in the same weekend, since weekends are relatively quiet for Homebrew/core.

Otherwise, we can wait till the next weekend. Or some other time Homebrew/core is quiet before then, but I suspect weekends are best.

**Comment by Bo98** [comment]
> That depends on when we can get the Intel Big Sur nodes up to 11.3

Doesn't technically need to be done first but I think it's probably likely to be anyway.

> I suggest we do that early this weekend, and then do the #11286 merge in the same weekend

That's fine by me. Keep me in the loop with how your testing is doing in the meantime.

**Comment by carlocab** [comment]
> That's fine by me. Keep me in the loop with how your testing is doing in the meantime.

Will do.

File: Library/Homebrew/extend/os/mac/keg_relocate.rb (hunk 1 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/mac/keg_relocate.rb:128]:
```suggestion
    rpaths = file.rpaths
                 .reject { |fn| fn =~ /^@(loader|executable)_path/ }
```

File: Library/Homebrew/extend/os/mac/keg_relocate.rb [hunk 2]
```diff
@@ -35,6 +35,18 @@ def relocate_dynamic_linkage(relocation)
...
+
+        if ENV["HOMEBREW_RELOCATE_RPATHS"]
+          each_rpath_for(file) do |old_name|
+            if old_name.start_with? relocation.old_cellar
+              new_name = old_name.sub(relocation.old_cellar, relocation.new_cellar)
+            elsif old_name.start_with? relocation.old_prefix
+              new_name = old_name.sub(relocation.old_prefix, relocation.new_prefix)
+            end
+
+            change_rpath(old_name, new_name, file) if new_name
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/mac/keg_relocate.rb:47]:
```suggestion
            new_name = if old_name.start_with? relocation.old_cellar
              old_name.sub(relocation.old_cellar, relocation.new_cellar)
            elsif old_name.start_with? relocation.old_prefix
              old_name.sub(relocation.old_prefix, relocation.new_prefix)
            else
              next
            end

            change_rpath(old_name, new_name, file)
```

**Review by MikeMcQuaid — DISMISSED** [review]
Looks good so far, some minor style tweaks. Would be good to have a test through this path if possible but np if not.

File: Library/Homebrew/extend/os/mac/keg_relocate.rb (hunk 1 above)
Comment (carlocab) [reply Library/Homebrew/extend/os/mac/keg_relocate.rb:128]: These style comments (including your other one) also apply to identical code just above these lines, though. Should I apply those changes there as well?

I suppose I could also refactor the duplication into a separate method. Hasn't passed the rule-of-three bar for that though.

**Comment by carlocab** [comment]
> Would be good to have a test through this path if possible but np if not.

Yep. It looks like all of `keg_relocate` is missing tests though. I'll try to create a proper one in a follow-up.

**Comment by carlocab** [comment]
FYI, I intend to merge this early this evening (before the end of the review period) so I can get a headstart on build-testing some formulae with `HOMEBREW_RELOCATE_RPATHS` (in an external tap).

File: Library/Homebrew/extend/os/mac/keg_relocate.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/extend/os/mac/keg_relocate.rb]: > These style comments (including your other one) also apply to identical code just above these lines, though. Should I apply those changes there as well?

Yes, ideally!



> I suppose I could also refactor the duplication into a separate method. Hasn't passed the rule-of-three bar for that though.

May be worth it regardless, though?

Comment (carlocab) [reply Library/Homebrew/extend/os/mac/keg_relocate.rb]: Cleaning up the other methods would make rebasing #11286 a bit messy, so I'll just work on them after that is merged. Left a note for myself: https://github.com/Homebrew/brew/pull/11286#issuecomment-833680029

**Pull request merged**